lib/pengine/utils.c

```diff
@@ -568,13 +568,41 @@ custom_action(pe_resource_t * rsc, char *key, const char *task,
         enum action_tasks a_task = text2task(action->task);
         enum pe_quorum_policy quorum_policy = effective_quorum_policy(rsc, data_set);
         int warn_level = LOG_TRACE;
+        xmlNode *op_entry = action->op_entry;
 
         if (save_action) {
             warn_level = LOG_WARNING;
         }
 
+        /* For a reload action, if no reload op is configured or the reload op
+         * doesn't have its own extra parameters, take start op's if there's
+         * any.
+         *
+         * A fact is, changes of start op's extra parameters result into a
+         * reload, and a successful reload gets updated as a start lrm_rsc_op
+         * based on pcmk__create_history_xml(). So this also ensures the digests
```

Comment:
I can't help but wonder if the most correct approach would be to record reload as reload, and handle it like start when unpacking history. Someone might want to trace just one of reload or start. (I'm not sure if there's a use case for action-specific instance attributes beyond trace_ra ...)

Reply:
AFAIK, OCF_CHECK_LEVEL is another op-specific parameter, which is normally only used  in monitor though.

I was also hesitant if it'd be worthy the effort to record reload as reload and take care of the unpacking of it,  which might be a further step towards the most correct way...  But I could think it'd still be a tricky situation in case that reload op has its own op-specific parameters. These questions keep occurring to me:

* Does it always make sense for changes of start-specific parameters trigger a reload?
* If so,  and if the reload just takes reload-specific parameters,  what's the point of doing reload?
* How does the unpacking part know whether recorded digests were previously calculated considering any reload-specific parameters or start-specific parameters?
...

I can tell any choice would introduce a chain reaction in terms of consolidating what parameters should be passed/considered/ruled out, how digests are recorded and calculated among the components ...

Reply:
Ugh, that does get messy. I think allowing arbitrary op-specific parameters is dangerous to begin with, though I don't see an obviously better approach.

One possibility would be to disallow reloads when start or stop has op-specific parameters, and always do a full stop and start in that case. It might be tricky to find the right place in code to do that, though.

BTW what agent is using reload in this case? I'm only aware of the ocf:pacemaker agents using reload in the pacemaker-specific sense. The new OCF 1.1 standard should help with that.

Reply:
> Ugh, that does get messy. I think allowing arbitrary op-specific parameters is dangerous to begin with, though I don't see an obviously better approach.
> 
> One possibility would be to disallow reloads when start or stop has op-specific parameters, and always do a full stop and start in that case. It might be tricky to find the right place in code to do that, though.

Good thinking. I'll take a look if there's a good place to achieve this.

> 
> BTW what agent is using reload in this case? I'm only aware of the ocf:pacemaker agents using reload in the pacemaker-specific sense. The new OCF 1.1 standard should help with that.

The issue was found by me with Dummy RA while I was asked to confirm the behaviors with trace_ra being added for different ops. Indeed I assume so far it doesn't particularly impact many resources in this regard. AFAICT, among the dozen of agents from resource-agents that support reload, probably only dovecot, mysql-proxy, named, nginx and postix tried to achieve sensible reload ..

Actually the most annoying thing I was trying to address was,  after addition of trace_ra parameter in start op triggered reload, from then on,  the resource would unexpectedly get reloaded in every single new transition as long as the start-specific parameter was existing, because the main digest would never match ...

Reply:
> The issue was found by me with Dummy RA while I was asked to confirm the behaviors with trace_ra being added for different ops. Indeed I assume so far it doesn't particularly impact many resources in this regard. AFAICT, among the dozen of agents from resource-agents that support reload, probably only dovecot, mysql-proxy, named, nginx and postix tried to achieve sensible reload ...

Ah that makes sense. None of the agents from resource-agents support reload as Pacemaker uses it, which is why the OCF 1.1 standard separates the reload and reload-agent actions. resource-agents will keep their reload action, which reloads the service's native configuration, and reload-agent will be used to reload agent parameters (as ocf:pacemaker agents currently use reload). So, Dummy may be the only thing affected. :)